packages/core-contracts/contracts/token/ERC20HistoricalBalance.sol

```diff
@@ -0,0 +1,135 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "../interfaces/IERC20HistoricalBalance.sol";
+import "./ERC20.sol";
+import "./ERC20HistoricalBalanceStorage.sol";
+
+/*
+    Reference implementations:
+    * OpenZeppelin - https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20Votes.sol
+*/
+
+contract ERC20HistoricalBalance is ERC20, IERC20HistoricalBalance, ERC20HistoricalBalanceStorage {
+    error InvalidUint32Number(uint blockNumber);
+    error BlockNumberNotYetMined(uint blockNumber);
+
+    function totalSupplyAt(uint blockNumber) external view override returns (uint) {
+        if (blockNumber >= block.number) {
+            revert BlockNumberNotYetMined(blockNumber);
+        }
+
+        return _checkpointsLookup(_erc20HistoricalBalanceStore().totalSupplyCheckpoints, blockNumber);
+    }
+
+    function balanceOfAt(address owner, uint blockNumber) public view override returns (uint) {
+        if (blockNumber >= block.number) {
+            revert BlockNumberNotYetMined(blockNumber);
+        }
+
+        return _checkpointsLookup(_erc20HistoricalBalanceStore().checkpoints[owner], blockNumber);
+    }
+
+    function _checkpointsLookup(Checkpoint[] storage ckpts, uint256 blockNumber) private view returns (uint256) {
+        // We run a binary search to look for the earliest checkpoint taken after `blockNumber`.
+        //
+        // During the loop, the index of the wanted checkpoint remains in the range [low-1, high).
+        // With each iteration, either `low` or `high` is moved towards the middle of the range to maintain the invariant.
+        // - If the middle checkpoint is after `blockNumber`, we look in [low, mid)
+        // - If the middle checkpoint is before or equal to `blockNumber`, we look in [mid+1, high)
+        // Once we reach a single value (when low == high), we've found the right checkpoint at the index high-1, if not
+        // out of bounds (in which case we're looking too far in the past and the result is 0).
+        // Note that if the latest checkpoint available is exactly for `blockNumber`, we end up with an index that is
+        // past the end of the array, so we technically don't find a checkpoint after `blockNumber`, but it works out
+        // the same.
+        uint256 high = ckpts.length;
+        uint256 low = 0;
+        while (low < high) {
+            uint256 mid = _average(low, high);
+            if (ckpts[mid].fromBlock > blockNumber) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        return high == 0 ? 0 : ckpts[high - 1].balance;
+    }
+
+    function _transfer(
```

Comment:
@ajsantander what is preferred here, overriding and calling super as I do here or adding a hook on ERC20 pre/post some actions (mint, burn, transfer) and override the hooks as done by OZs? 
[OZ solution](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/b3b83b558ebb9982e27ae5ee0bb5f33f278863dd/contracts/token/ERC20/ERC20.sol#L248)

Reply:
I think overrides are more explicit, but don't have a strong opinion.